File: src/main/java/it/unibo/crabinv/View/PauseMenu.java
package it.unibo.crabinv.View;

import it.unibo.crabinv.Controller.core.audio.AudioController;
import it.unibo.crabinv.Controller.core.i18n.LocalizationController;
import it.unibo.crabinv.Model.core.audio.SFXTracks;
import it.unibo.crabinv.Model.core.i18n.TextKeys;
import it.unibo.crabinv.SceneManager;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Button;
import javafx.scene.control.CheckBox;
import javafx.scene.control.Label;
import javafx.scene.layout.*;

/**
 * Provides the method to create the pause menu GUI
 */
public class PauseMenu {
    private final SceneManager sceneManager;
    private final LocalizationController loc;
    private final AudioController audio;
    private final Runnable resumeMethod;
    private final Pos settingsAlignment = Pos.CENTER_LEFT;
    private final SettingComponentsGenerator components;

    /**
     * @param sceneManager the instance of SceneManager
     * @param loc the instance of LocalizationController
     * @param audio the instance of AudioController
     */
    public PauseMenu(SceneManager sceneManager, LocalizationController loc, AudioController audio, Runnable resumeMethod) {
        this.sceneManager = sceneManager;
        this.loc = loc;
        this.audio = audio;
        this.resumeMethod = resumeMethod;
        components = new SettingComponentsGenerator(sceneManager, loc, audio, settingsAlignment);
    }

    /**
     * @return the Stackpane containing a working pauseMenu
     */
    public Pane getView() {
        StackPane pane = new StackPane();
        Label title = new Label(loc.getString(TextKeys.PAUSE));
        title.getStyleClass().add("title");
        StackPane.setAlignment(title, Pos.TOP_CENTER);
        StackPane.setMargin(title, new Insets(20,0,0,0));
        pane.getChildren().add(title);
        VBox mainColumn = new VBox(25);
        mainColumn.setMinWidth(350);
        GridPane grid = new GridPane();
        grid.addColumn(0);
        grid.addColumn(1, mainColumn);
        grid.addColumn(2);
        grid.setAlignment(Pos.CENTER);
        mainColumn.setAlignment(settingsAlignment);
        HBox bgmVolume = components.createVolumeSlider(audio.getBGMVolume(), audio::setBGMVolume, audio::playSFX, TextKeys.BGM_VOLUME);
        HBox sfxVolume = components.createVolumeSlider(audio.getSFXVolume(), audio::setSFXVolume, audio::playSFX, TextKeys.SFX_VOLUME);
        CheckBox bgmMute = components.createMute(TextKeys.BGM_MUTE, audio.isBGMMuted(), audio::toggleBGMMute);
        CheckBox sfxMute = components.createMute(TextKeys.SFX_MUTE, audio.isSFXMuted(), audio::toggleSFXMute);
        Button exit = new Button(loc.getString(TextKeys.EXIT_GAME));
        exit.setOnAction(_ -> {
            sceneManager.showMainMenu();
            audio.playSFX(SFXTracks.MENU_SELECT);
        });
        exit.getStyleClass().add("app-button");
        StackPane.setAlignment(exit, Pos.BOTTOM_CENTER);
        Button resume = new Button(loc.getString(TextKeys.RESUME));
        resume.setOnAction(_ -> {
            sceneManager.hidePauseMenu();
            resumeMethod.run();
            audio.playSFX(SFXTracks.MENU_SELECT);
        });
        StackPane.setMargin(exit, new Insets(0,0,60,0));
        mainColumn.getChildren().addAll(bgmVolume, sfxVolume, bgmMute, sfxMute);
        for (var child : mainColumn.getChildren()) {
            child.focusedProperty().addListener(_ -> audio.playSFX(SFXTracks.MENU_HOVER));
        }
        pane.getChildren().addAll(grid, exit, resume);
        return pane;
    }
}

File: src/main/java/it/unibo/crabinv/view/PauseMenu.kt
package it.unibo.crabinv.view

import it.unibo.crabinv.SceneManager
import it.unibo.crabinv.controller.core.audio.AudioController
import it.unibo.crabinv.controller.core.i18n.LocalizationController
import it.unibo.crabinv.model.core.audio.SFXTracks
import it.unibo.crabinv.model.core.i18n.TextKeys
import javafx.beans.Observable
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Pos
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.layout.HBox
import javafx.scene.layout.Pane
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox

/**
 * Provides the method to create the pause menu GUI.
 * @param loc the instance of LocalizationController
 * @param audio the instance of AudioController
 * @param resumeMethod the method that calls the resume of the game engine
 * @param gameOver the method that calls the gameOver of the game engine
 */
class PauseMenu(
    private val sceneManager: SceneManager,
    private val loc: LocalizationController,
    private val audio: AudioController,
    private val resumeMethod: Runnable,
    private val gameOver: Runnable,
) {
    private val components: SettingsComponentsGenerator =
        SettingsComponentsGenerator(sceneManager, loc, audio, ViewParameters.SETTINGS_ALIGMENT)

    val view: Pane
        /**
         * @return the Stackpane containing a working pauseMenu
         */
        get() {
            val pane = StackPane()
            pane.styleClass.add("pause-pane")
            val content = VBox(ViewParameters.DEFAULT_SPACING.toDouble())
            content.alignment = Pos.CENTER
            content.maxWidth = ViewParameters.DEFAULT_PAUSE_WIDTH.toDouble()
            val title = Label(loc.getString(TextKeys.PAUSE))
            title.styleClass.add("title")
            val mainColumn =
                VBox(ViewParameters.DEFAULT_SPACING.toDouble())
            mainColumn.alignment = ViewParameters.SETTINGS_ALIGMENT
            val bgmVolume =
                components.createVolumeSlider(
                    audio.bgmVolume,
                    { volume: Int -> audio.bgmVolume = volume },
                    { sfx: SFXTracks? -> audio.playSFX(sfx!!) },
                    TextKeys.BGM_VOLUME,
                )
            val sfxVolume =
                components.createVolumeSlider(
                    audio.sfxVolume,
                    { volume: Int -> audio.sfxVolume = volume },
                    { sfx: SFXTracks? -> audio.playSFX(sfx!!) },
                    TextKeys.SFX_VOLUME,
                )
            val bgmMute =
                components.createMute(
                    TextKeys.BGM_MUTE,
                    audio.isBGMMuted,
                ) { audio.toggleBGMMute() }
            val sfxMute =
                components.createMute(
                    TextKeys.SFX_MUTE,
                    audio.isSFXMuted,
                ) { audio.toggleSFXMute() }
            mainColumn.children.addAll(bgmVolume, sfxVolume, bgmMute, sfxMute)
            for (child in mainColumn.children) {
                child
                    .focusedProperty()
                    .addListener { `_`: Observable? ->
                        audio.playSFX(SFXTracks.MENU_HOVER)
                    }
            }
            val resume =
                createPauseMenuButton(
                    loc.getString(TextKeys.RESUME),
                    resumeMethod,
                ) { sceneManager.hidePauseMenu() }
            val exit =
                createPauseMenuButton(
                    loc.getString(TextKeys.ABANDON),
                    gameOver,
                ) { sceneManager.showMainMenu() }
            val buttons =
                HBox(ViewParameters.DEFAULT_SPACING.toDouble(), resume, exit)
            buttons.alignment = Pos.CENTER
            content.children.addAll(title, mainColumn, buttons)
            pane.children.add(content)
            return pane
        }

    /**
     * Creates the buttons for the pause menu.
     * @param text the text of the button
     * @param method the method it's linked to
     * @param changeScene the method to change the scene
     * @return the complete button
     */
    private fun createPauseMenuButton(
        text: String?,
        method: Runnable,
        changeScene: Runnable,
    ): Button {
        val button = Button(text)
        button.styleClass.add("app-button")
        button.onAction =
            EventHandler { `_`: ActionEvent? ->
                method.run()
                changeScene.run()
                audio.playSFX(SFXTracks.MENU_SELECT)
            }
        button
            .focusedProperty()
            .addListener { `_`: Observable? -> audio.playSFX(SFXTracks.MENU_HOVER) }
        return button
    }
}
